src/aes.rs

```diff
@@ -0,0 +1,276 @@
+        rng.fill_bytes(&mut buffer);
+        buffer
+    };
+    let (iv, edek) = aes_encrypt(kek, &document_key, &[], rng)?;
+    let aes_edek = icl_header_v4::v4document_header::edek_wrapper::Aes256GcmEncryptedDek {
+        ciphertext: edek.0.into(),
+        iv: Bytes::copy_from_slice(&iv),
+        id: id.into(),
+        ..Default::default()
+    };
+    Ok((EncryptionKey(document_key), aes_edek))
+}
+
+/// Generate an aes edek, and encrypt it using the kek. The provided id will be put into the Aes256GcmEdek.
+/// The edek will be placed into a V4DocumentHeader and the signature will be computed.
+/// The generated aes dek is the key used to compute the signature.
+pub fn generate_aes_edek_and_sign<R: CryptoRng + RngCore>(
+    rng: &mut R,
+    kek: EncryptionKey,
+    id: &str,
+) -> Result<(EncryptionKey, icl_header_v4::V4DocumentHeader)> {
+    let (aes_dek, aes_edek) = generate_aes_edek(rng, kek, id)?;
+    Ok((
+        aes_dek,
+        create_signed_header(
+            icl_header_v4::v4document_header::EdekWrapper {
+                edek: Some(Edek::Aes256GcmEdek(aes_edek)),
+                ..Default::default()
+            },
+            aes_dek,
+        ),
+    ))
+}
+
+/// Decrypt the aes edek. Does not verify signature of the header or check that the id is appropriate.
+/// You must do that as a separate step.
+pub fn decrypt_aes_edek(
+    kek: &EncryptionKey,
+    aes_edek: &icl_header_v4::v4document_header::edek_wrapper::Aes256GcmEncryptedDek,
+) -> Result<EncryptionKey> {
+    let iv = aes_edek.iv.as_ref().try_into().map_err(|_| {
+        Error::DecryptError("IV from the edek was not the correct length.".to_string())
+    })?;
+    aes_decrypt(kek, iv, &aes_edek.ciphertext, &[])
+        .and_then(|dek_bytes| {
+            dek_bytes
+                .try_into()
+                .map_err(|_| Error::DecryptError("iv was not of the correct size".to_string()))
+        })
+        .map(|dek_bytes| EncryptionKey(dek_bytes))
+}
+
+/// Decrypt a V4 detatched document. The document should have the expected header
+pub fn decrypt_detatched_document(
+    key: &EncryptionKey,
+    payload: EncryptedPayload,
+) -> Result<PlaintextDocument> {
+    let payload_len = payload.0.len();
+    if payload_len < DETATCHED_HEADER_LEN + IV_LEN {
+        Err(Error::EdocTooShort(payload_len))
+    } else {
+        let header = payload.0.slice(0..DETATCHED_HEADER_LEN);
+        if header != [&[V0], &MAGIC[..]].concat() {
+            Err(Error::NoIronCoreMagic)
+        } else {
+            let iv = payload
+                .0
+                .slice(DETATCHED_HEADER_LEN..(DETATCHED_HEADER_LEN + IV_LEN))
+                .as_ref()
+                .try_into()
+                .expect("IV conversion will always have 12 bytes.");
+            let ciphertext = payload.0.slice(DETATCHED_HEADER_LEN + IV_LEN..);
+            aes_decrypt(key, iv, &ciphertext[..], &[]).map(|v| PlaintextDocument(v))
+        }
+    }
```

Comment:
How would you feel about something like this? I think it's easier to parse than the `slice`s.
```suggestion
    let payload_len = payload.0.len();
    if payload_len < DETATCHED_HEADER_LEN + IV_LEN {
        Err(Error::EdocTooShort(payload_len))
    } else {
        let (header, iv_and_cipher) = payload.0.split_at(DETATCHED_HEADER_LEN);
        if header != [&[V0], &MAGIC[..]].concat() {
            Err(Error::NoIronCoreMagic)
        } else {
            let (iv_slice, ciphertext) = iv_and_cipher.split_at(IV_LEN);
            let iv = iv_slice
                .try_into()
                .expect("IV conversion will always have 12 bytes.");
            aes_decrypt(key, iv, &ciphertext, &[]).map(PlaintextDocument)
        }
    }
```

Reply:
I think that's fine. I don't find it much easier, but I can see how some people might so for clarity it's probably worth it.